File: go/adapters/exec/exec.go
package exec

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"

	xrr "hop.top/xrr"
	"gopkg.in/yaml.v3"
)

// Request represents an exec interaction request.
//
// Cwd is the working directory the subprocess was launched from. When
// non-empty it participates in the fingerprint, so the same command run
// in different directories produces distinct cassette keys — essential
// for cross-process e2e adopters (see XRR_CASSETTE_DIR pattern) whose
// tests invoke the same binary many times under different cwds within a
// single parent cassette dir.
//
// This is a Go-only extension to the v1 cassette spec. Within the Go
// port it is backward compatible: leaving Cwd empty preserves the
// canonical argv+stdin fingerprint and cassettes recorded before this
// field existed still match. But cassettes recorded with NON-EMPTY Cwd
// will NOT replay in ts / py / rs / php ports until those ports adopt
// the same rule — their fingerprint calculation will miss. Use
// non-empty Cwd only when record and replay happen in runtimes that
// agree on the extension, or leave Cwd empty to preserve cross-runtime
// replay. See spec/cassette-format-v1.md "Exec Fingerprint Inputs" for
// the formal status of this extension.
type Request struct {
	Argv  []string          `yaml:"argv"  json:"argv"`
	Stdin string            `yaml:"stdin,omitempty" json:"stdin,omitempty"`
	Cwd   string            `yaml:"cwd,omitempty"   json:"cwd,omitempty"`
	Env   map[string]string `yaml:"env,omitempty"   json:"env,omitempty"`
}

func (r *Request) AdapterID() string { return "exec" }

// Response represents an exec interaction response.
type Response struct {
	Stdout     string `yaml:"stdout"`
	Stderr     string `yaml:"stderr,omitempty"`
	ExitCode   int    `yaml:"exit_code"`
	DurationMs int64  `yaml:"duration_ms,omitempty"`
}

func (r *Response) AdapterID() string { return "exec" }

// Adapter implements xrr.Adapter for exec interactions.
type Adapter struct{}

// NewAdapter returns a new exec Adapter.
func NewAdapter() *Adapter { return &Adapter{} }

func (a *Adapter) ID() string { return "exec" }

// Fingerprint returns sha256(canonical JSON of {argv, stdin, cwd?})[:8].
//
// cwd only participates in the hash when non-empty. This keeps
// backwards compatibility: adopters that don't populate Request.Cwd
// get the legacy argv+stdin-only fingerprint, so cassettes recorded
// before this field existed still match.
func (a *Adapter) Fingerprint(req xrr.Request) (string, error) {
	r, ok := req.(*Request)
	if !ok {
		return "", fmt.Errorf("exec: unexpected request type %T", req)
	}
	fields := map[string]any{
		"argv":  r.Argv,
		"stdin": r.Stdin,
	}
	if r.Cwd != "" {
		fields["cwd"] = r.Cwd
	}
	canonical, err := json.Marshal(fields)
	if err != nil {
		return "", fmt.Errorf("exec: fingerprint marshal: %w", err)
	}
	sum := sha256.Sum256(canonical)
	return fmt.Sprintf("%x", sum[:4]), nil // 4 bytes = 8 hex chars
}

// Serialize marshals v as YAML.
func (a *Adapter) Serialize(v any) ([]byte, error) {
	return yaml.Marshal(v)
}

// Deserialize unmarshals data into target.
func (a *Adapter) Deserialize(data []byte, target any) error {
	return yaml.Unmarshal(data, target)
}

File: py/src/xrr/adapters/exec.py
"""exec adapter — fingerprints on argv + stdin."""
from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass, field
from typing import Any


@dataclass
class ExecRequest:
    argv: list[str]
    stdin: str = ""
    env: dict[str, str] = field(default_factory=dict)


@dataclass
class ExecResponse:
    stdout: str
    stderr: str = ""
    exit_code: int = 0
    duration_ms: int = 0


class ExecAdapter:
    id = "exec"

    def fingerprint(self, req: ExecRequest) -> str:
        key = {"argv": req.argv, "stdin": req.stdin}
        canonical = json.dumps(key, sort_keys=True, separators=(",", ":"))
        return hashlib.sha256(canonical.encode()).hexdigest()[:8]

    def serialize_req(self, req: ExecRequest) -> dict[str, Any]:
        return {"argv": req.argv, "stdin": req.stdin, "env": req.env}

    def serialize_resp(self, resp: ExecResponse) -> dict[str, Any]:
        return {
            "stdout": resp.stdout,
            "stderr": resp.stderr,
            "exit_code": resp.exit_code,
            "duration_ms": resp.duration_ms,
        }

    def deserialize_req(self, data: dict[str, Any]) -> ExecRequest:
        return ExecRequest(
            argv=data["argv"],
            stdin=data.get("stdin", ""),
            env=data.get("env", {}),
        )

    def deserialize_resp(self, data: dict[str, Any]) -> ExecResponse:
        return ExecResponse(
            stdout=data.get("stdout", ""),
            stderr=data.get("stderr", ""),
            exit_code=data.get("exit_code", 0),
            duration_ms=data.get("duration_ms", 0),
        )
